{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd3911ad-a065-46f7-bfaf-05ad3c7c6bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pdb\n",
    "import numpy as np\n",
    "import itertools\n",
    "from snakemake.io import expand\n",
    "import yaml\n",
    "import os\n",
    "\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "30cabd11-d61c-44e1-a2e4-24b61b728b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = 'config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc360584-28e8-4de8-958c-69ca9510c24b",
   "metadata": {},
   "source": [
    "## 231017 pipeline restructuring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e2d79393-02ee-4413-a8e0-5d8fd873bebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "configfile: 'config.yml'\n",
    "config_tsv = '231018_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d94e3529-64e2-4660-a01d-6dcbf019e08d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC',\n",
    "      'biorep_num': '1',\n",
    "      'flowcell': '1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8fc0bf4f-8895-47be-bea7-d254b93fe74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subset_df_on_wcs(wc, df):\n",
    "    \"\"\"\n",
    "    Return a copy of the input metadata df limited to the wildcards\n",
    "    \"\"\"\n",
    "    temp = df.copy(deep=True)\n",
    "    for key, item in wc.items():\n",
    "        if type(item) == list:\n",
    "            temp = temp.loc[temp[key].isin(item)]\n",
    "        else:\n",
    "            temp = temp.loc[temp[key] == item]\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6706ce46-5618-4532-bc6f-1542b1c6132e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cfg_entries(wc, df, cfg_entry):\n",
    "    \"\"\"\n",
    "    Expand a config entry based on the wildcards and the \n",
    "    values in df that satisfy these wildcards\n",
    "    \"\"\"\n",
    "    temp = subset_df_on_wcs(wc, df)\n",
    "    study = temp.study.tolist()\n",
    "    genotype = temp.genotype.tolist()\n",
    "    sex = temp.sex.tolist()\n",
    "    age = temp.age.tolist()\n",
    "    tissue = temp.tissue.tolist()\n",
    "    biorep_num = temp.biorep_num.tolist()\n",
    "    flowcell = temp.flowcell.tolist()\n",
    "\n",
    "    files = expand(cfg_entry,\n",
    "                   zip,\n",
    "                   study=study,\n",
    "                   genotype=genotype,\n",
    "                   sex=sex,\n",
    "                   age=age,\n",
    "                   tissue=tissue,\n",
    "                   biorep_num=biorep_num,\n",
    "                   flowcell=flowcell)\n",
    "    return files\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f6c71d80-16c6-4ca5-9843-ed2aefd56232",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_col(wc, df, col):\n",
    "    \"\"\"\n",
    "    From the metadata dataframe df, get the entries that satisfy\n",
    "    the wildcards requirements and return the corresponding value\n",
    "    from col. Ensure that this is always a 1:1 relationship, otherwise\n",
    "    throw an error.\n",
    "    \"\"\"\n",
    "    cols = [col] + [key for key, item in wc.items()]\n",
    "    # for key, item in wc.items():\n",
    "    #     if type(item) == list:\n",
    "    #         temp = temp.loc[temp[key].isin(item)]\n",
    "    #     else:\n",
    "    #         temp = temp.loc[temp[key] == item]\n",
    "\n",
    "    temp = subset_df_on_wcs(wc, df)\n",
    "    temp = temp[cols].drop_duplicates()\n",
    "    \n",
    "    if len(temp.index) != 1:\n",
    "        msg = 'Issues getting data from DF with wildcards'\n",
    "        for key, item in wc.items():\n",
    "            msg+=f'\\n{key}: {item}'\n",
    "        raise ValueError(msg)\n",
    "\n",
    "    val = temp[col].tolist()[0]\n",
    "    return val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d66fda86-8a38-469a-a5eb-a10f37ceff0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/talon_label/5xFADHEMI_F_4_months_HC_1_1_labeled_sorted.bam',\n",
       " 'data/talon_label/5xFADHEMI_F_4_months_HC_1_2_labeled_sorted.bam']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfg_entry = config['talon_label']['sort_bam']\n",
    "wc = {'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC',\n",
    "      'biorep_num': '1'}\n",
    "\n",
    "temp = subset_df_on_wcs(wc, df)\n",
    "study = temp.study.tolist()\n",
    "genotype = temp.genotype.tolist()\n",
    "sex = temp.sex.tolist()\n",
    "age = temp.age.tolist()\n",
    "tissue = temp.tissue.tolist()\n",
    "biorep_num = temp.biorep_num.tolist()\n",
    "flowcell = temp.flowcell.tolist()\n",
    "\n",
    "files = expand(cfg_entry,\n",
    "               zip,\n",
    "               study=study,\n",
    "               genotype=genotype,\n",
    "               sex=sex,\n",
    "               age=age,\n",
    "               tissue=tissue,\n",
    "               biorep_num=biorep_num,\n",
    "               flowcell=flowcell)\n",
    "return files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "40bee402-6f45-4988-9b1c-7e27aadea48f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "1  ad003_11616_lig-blk_2_t1.fastq   \n",
       "2  ad003_11617_lig-blk_1_t1.fastq   \n",
       "3  ad003_11617_lig-blk_2_t1.fastq   \n",
       "4  ad003_11625_lig-blk_1_t2.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "1  ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "2  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "3  ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "4  ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "1   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "2   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "3   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "4   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "\n",
       "                     dataset  \n",
       "0  5xFADHEMI_F_4_months_HC_1  \n",
       "1  5xFADHEMI_F_4_months_HC_1  \n",
       "2  5xFADHEMI_F_4_months_HC_2  \n",
       "3  5xFADHEMI_F_4_months_HC_2  \n",
       "4  5xFADHEMI_F_4_months_HC_3  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "911dbea0-984a-4ed9-8d54-a8e7945e470c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/share/crsp/lab/seyedam/share/ad_nanopore_tmp/AD003/ad003_trimfastq/ad003_11616_lig-blk_1_t1.fastq'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_df_col(wc, df, 'fname')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3ef13cc8-043e-4ecb-baab-9ed4468af31a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replicates in modelad consist of \n",
    "# * genotype\n",
    "# * sex\n",
    "# * age\n",
    "# * tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a1c223-7da3-41b1-9a2f-9741ef7f02e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
